addons/pos_sale/static/src/js/models.js

```diff
@@ -83,7 +83,7 @@ models.Orderline = models.Orderline.extend({
    * @param {'sale.order.line'} saleOrderLine
    */
   setQuantityFromSOL: function(saleOrderLine) {
-      if (this.product.type === 'service') {
+      if (this.product.type === 'service' && this.product.invoice_policy !== 'order') {
```

Comment:
Sorry for the ping-pong of ideas. Now, I think the main issue is about the "state" of the sale order. Since it is not yet "sale/done", the `qty_to_invoice` is always zero (check its getter). And since "service" products relies on it, pos tries to settle the wrong quantity.

In fact, what we should do is to replicate the result of the `qty_to_invoice` getter such that it is also applicable for "quotations".

https://github.com/odoo/odoo/blob/e4e9a14cc9f8d752a29ee07ca7f6daadf5b93b34/addons/sale/models/sale_order_line.py#L80

We probably need to modify the `qty_to_invoice` field inside the `read_converted` method. 

https://github.com/odoo/odoo/blob/e4e9a14cc9f8d752a29ee07ca7f6daadf5b93b34/addons/pos_sale/models/sale_order.py#L69

**With regards the invoice policy solution**

`invoice_policy` affects the calculation of `qty_to_invoice` only when the sale order is validated (not in draft). The solution you propose seems to be correct by luck because for `qty_to_invoice` for `delivered` quantities, the calculation is `product_uom_qty - qty_delivered`.

We could even do the fix here by if we have the state of the order:

```
if (this.product.type === 'service' && saleOrder.state != 'draft') {
  this.set_quantity(saleOrderLine.qty_to_invoice);
} else {
  this.set_quantity(saleOrderLine.product_uom_qty - Math.max(saleOrderLine.qty_delivered, saleOrderLine.qty_invoiced));
}
```

If I remember right, that's your previous solution. I'm really sorry for the back-and-forth.

Reply:
No problem at all ! 

I also included the state `sent`, as for "Quotation sent' to the condition. This is the state of a quotation sent through the website that hasn't been paid yet, using the pay in store option for example. 

I made a quick test. Using the same abbreviations as before here's how it went. 

I made a quotation with 3 PSD and 2 PSO and saved it. 
* If I choose to settle it, quantities are PSD = 3 and PSO = 2
* If I make a down payment, all quantities are taken into account and the order is now a sale, not a quotation anymore
  * From there if I settle the rest without making delivery I will have PSD=0, PSO = 2, down payment -1
  * I deliver 2 out of 3 products
  * I settle again, PSD= 2, PSO=0, down payment -1, down payment 1 (since already taken into account at first settle